Source/Renderer/GLSLModernizer.js

```diff
@@ -0,0 +1,233 @@
+     *
+     * This function is nowhere near comprehensive or complete. It just
+     * handles some common cases.
+     *
+     * Note that this function requires the presence of the
+     * "#define OUTPUT_DECLARATION" line that is appended
+     * by ShaderSource.
+     *
+     * @private
+     */
+    function glslModernizeShaderText(source, isFragmentShader) {
+        var outputDeclarationRegex = /#define OUTPUT_DECLARATION/;
+        var splitSource = source.split('\n');
+
+        if (source.search(/#version 300 es/g) !== -1) {
+            return source;
+        }
+
+        var outputDeclarationLine = -1;
+        var number;
+        for (number = 0; number < splitSource.length; ++number) {
+            var line = splitSource[number];
+            if (outputDeclarationRegex.exec(line)) {
+                outputDeclarationLine = number;
+                break;
+            }
+        }
+
+        if (outputDeclarationLine === -1) {
+            throw new DeveloperError('Could not find a #define OUTPUT_DECLARATION nor a main function!');
+        }
+
+        // Note that this fails if your string looks like
+        // searchString[singleCharacter]searchString
+        function replaceInSource(str, replacement, strIsRegex) {
+            var regex = str;
+            if ((strIsRegex === undefined) || strIsRegex === false) {
+                var regexStr = '(^|[^\\w])(' + str + ')($|[^\\w])';
+                regex = new RegExp(regexStr, 'g');
+            }
+
+            for (var number = 0; number < splitSource.length; ++number) {
+                var line = splitSource[number];
+                if (strIsRegex) {
+                    splitSource[number] = line.replace(regex, replacement);
+                } else {
+                    splitSource[number] = line.replace(regex, '$1' + replacement + '$3');
+                }
+            }
+        }
+
+        function findInSource(str) {
+            var regexStr = '(^|[^\\w])(' + str + ')($|[^\\w])';
+            var regex = new RegExp(regexStr, 'g');
+            for (var number = 0; number < splitSource.length; ++number) {
+                var line = splitSource[number];
+                if (regex.test(line)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        function compileSource() {
+            var wholeSource = '';
+            for (var number = 0; number < splitSource.length; ++number) {
+                wholeSource += splitSource[number] + '\n';
+            }
+            return wholeSource;
+        }
+
+        function setAdd(variable, set1) {
+            if (set1.indexOf(variable) === -1)
+                {set1.push(variable);}
+        }
+
+        function getVariablePreprocessorBranch(variablesThatWeCareAbout) {
+            var variableMap = {};
+            var negativeMap = {};
+
+            for (var a = 0; a < variablesThatWeCareAbout.length; ++a) {
```

Comment:
Throughout, store `length` as a separate variable, except those areas where length is dynamic.